flow/shared/authenticator.go

```diff
@@ -0,0 +1,218 @@
+	Enabled       bool   `json:"enabled" yaml:"enabled" mapstructure:"enabled"`
+}
+
+type identityProvider struct {
+	keySet      jwk.Set
+	validateOpt jwt.ValidateOption
+	issuer      string
+}
+
+func AuthGrpcMiddleware() ([]grpc.ServerOption, error) {
+	devKeySetJSON, _ := os.LookupEnv("PEERDB_OAUTH_KEY_SET_JSON")
+	auth0Domain, _ := os.LookupEnv("PEERDB_OAUTH_AUTH0DOMAIN")
+	auth0ClientID, _ := os.LookupEnv("PEERDB_OAUTH_AUTH0CLIENTID")
+	cfg := AuthenticationConfig{
+		Enabled:       devKeySetJSON != "" || auth0Domain != "",
+		DevKeySetJSON: devKeySetJSON,
+		Auth0Domain:   auth0Domain,
+		Auth0ClientID: auth0ClientID,
+	}
+	// load identity providers before checking if authentication is enabled so configuration can be validated
+	ip, err := identityProvidersFromConfig(cfg)
+
+	if !cfg.Enabled {
+		if err != nil { // if there was an error loading identity providers, warn only if authentication is disabled
+			slog.Warn("failed to initialize JWK key set", slog.Any("error", err))
+		}
+
+		slog.Warn("authentication is disabled")
+
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return []grpc.ServerOption{
+		grpc.ChainUnaryInterceptor(func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+			var authHeader string
+			authHeaders := metadata.ValueFromIncomingContext(ctx, "authorization")
+			if len(authHeaders) == 1 {
+				authHeader = authHeaders[0]
+			} else if len(authHeaders) > 1 {
+				return nil, errors.New("multiple Authorization headers supplied, request rejected")
+			}
+			_, err := validateRequestToken(authHeader, ip...)
+			if err != nil {
+				slog.Debug("failed to validate request token", slog.Any("error", err))
+				return nil, err
+			}
+			return handler(ctx, req)
+		}),
+	}, nil
+}
+
+func validateRequestToken(authHeader string, ip ...identityProvider) ([]byte, error) {
+	payload, err := jwtFromRequest(authHeader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse authorization header: %w", err)
+	}
+
+	token, err := jwt.ParseInsecure(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse token: %w", err)
+	}
+
+	provider, err := identityProviderByToken(ip, token)
+	if err != nil {
+		return nil, err
+	}
+
+	validateOpts := identityProviderValidateOpts(provider)
+	if err := jwt.Validate(token, validateOpts...); err != nil {
+		return nil, fmt.Errorf("failed to validate token: %w", err)
+	}
+
+	if _, err := jws.Verify(payload, jws.WithKeySet(provider.keySet)); err != nil {
+		return nil, fmt.Errorf("failed to verify token: %w", err)
+	}
+
+	return payload, nil
+}
+
+// jwtFromRequest extracts the JWT token from the Authorization header.
+// it truncates the "Bearer" prefix from the header value if exists.
+func jwtFromRequest(authHeader string) ([]byte, error) {
+	if authHeader == "" {
+		return nil, errors.New("missing Authorization header")
+	}
+
+	if len(authHeader) > 7 && authHeader[:7] == "Bearer " {
```

Comment:
how about https://pkg.go.dev/strings#TrimPrefix

Reply:
Can look to get that implemented upstream